File: 06-linked_list/00206-Reverse_Linked_List/00206-reverse_linked_list.cpp
// 206: Reverse Linked List
// https://leetcode.com/problems/reverse-linked-list/


#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

void llInput(ListNode** head, vector<int> inputs) {
    for (int i=inputs.size()-1; i>=0; i--) {
        ListNode* node = new ListNode(inputs[i], *head);
        *head = node;
    }
}

void llOutput(ListNode* head) {
    while (head != NULL) {
        cout << head->val << " -> ";
        head = head->next;
    }
    cout << "NULL" << endl;
}



class Solution {
public:
    // SOLUTION
    ListNode* reverseList(ListNode* head) {
        if (head == NULL || head->next == NULL) return head;
        
        ListNode* previous = NULL;
        ListNode* current = head;
        ListNode* next = current->next;

        while (current != NULL) {
            next = current->next;
            current->next = previous;
            previous = current;
            current = next;
        }

        return previous;
    }
};



int main() {
    Solution o;
    ListNode* head = NULL;

    // INPUT
    vector<int> li = {1,2,3,4,5};
    llInput(&head, li);

    // OUTPUT
    auto result = o.reverseList(head);
    llOutput(result);

    return 0;
}

File: 06-linked_list/00206-Reverse_Linked_List/00206-reverse_linked_list.py
# 206: Reverse Linked List
# https://leetcode.com/problems/reverse-linked-list/


class ListNode:
    def __init__(self, x: int):
        self.val = x
        self.next = None

class LinkedList:
    def __init__(self) -> None:
        self.head = None

    def Input(self, inputs: list[int]):
        for i in inputs:
            node = ListNode(i)
            if self.head:
                tail = self.head
                while tail.next != None: tail = tail.next
                tail.next = node
            else:
                self.head = node
    
    def Output(self, head: ListNode):
        while (head != None):
            print(head.val, "-> ", end='')
            head = head.next
        print("NULL")


class Solution:
    # SOLUTION
    def reverseList(self, head: ListNode) -> ListNode:
        if head == None or head.next == None: return head
        
        previous = None
        current = head
        next = current.next

        while (current != None):
            next = current.next
            current.next = previous
            previous = current
            current = next

        return previous


if __name__ == "__main__":
    o = Solution()
    ll = LinkedList()

    # INPUT
    li = [1,2,3,4,5]
    ll.Input(li)

    # OUTPUT
    result = o.reverseList(ll.head)
    ll.Output(result)
